File: KEYBOARD_ORBY_V3/RASPBERRY_PROJECT/ORBY_V4/OrbyGUI/electron/macros.js
const { shell, clipboard } = require('electron');
const { exec } = require('child_process');
const config = require('./config');
const plugins = require('./plugins');

// Ejecuta en el PC las macros que el propio teclado NO puede reproducir solo
// (avisa por CDC con MACRO:<id> en vez de tocarlas él, ver trigger_macro en
// main.cpp): hoy eso es únicamente lo que necesita posición ABSOLUTA del
// ratón, que exige conocer la resolución/escala de la pantalla. El resto de
// pasos (tecla, espera, clic, movimiento relativo) los toca el teclado por su
// cuenta —más rápido y sin depender de que esta app esté abierta— así que ya
// no deberían llegar aquí; se deja el reproductor completo por compatibilidad
// con macros antiguas que no se hayan podido subir al dispositivo.

let nutjs = null;
let nutjsFailed = false;

function loadNutJs() {
  if (nutjs || nutjsFailed) return nutjs;
  try {
    nutjs = require('@nut-tree-fork/nut-js');
    // nut.js mete su propio hueco ANTES de cada pulsación/clic (keyboard 300ms,
    // mouse 100ms, ver keyboard.class.js/mouse.class.js autoDelayMs), pensado
    // para separar llamadas sueltas hechas a mano. Aquí ya controlamos nosotros
    // el ritmo -la "Espera" entre pasos y el "gap" entre repeticiones, ver
    // repeatWithGap-, así que ese hueco extra solo suma un despropósito por
    // pulsación (con 11 repeticiones, 11 × 600ms = 6.6s de nada) que ningún
    // ajuste del editor podía tocar: no era el gap configurable, era este.
    nutjs.keyboard.config.autoDelayMs = 0;
    nutjs.mouse.config.autoDelayMs = 0;
  } catch (err) {
    nutjsFailed = true;
    console.error('No se pudo cargar @nut-tree-fork/nut-js:', err.message);
  }
  return nutjs;
}

// Solo cubre lo que se puede grabar desde el editor de la app (letras, dígitos,
// Enter, Escape, Espacio). Para ampliarlo basta con añadir la entrada aquí: el
// editor ya captura cualquier KeyboardEvent.code, solo falta traducirlo.
function mapCodeToNutKey(code, Key) {
  if (/^Key[A-Z]$/.test(code)) return Key[code.slice(3)];
  if (/^Digit[0-9]$/.test(code)) return Key[`Num${code.slice(5)}`];
  if (code === 'Enter') return Key.Enter;
  if (code === 'Escape') return Key.Escape;
  if (code === 'Space') return Key.Space;
  return null;
}

const CLICK_BUTTONS = { left: 'LEFT', middle: 'MIDDLE', right: 'RIGHT' };

// Acciones de energía: no hay opcode de firmware para esto (igual que
// open_app), siempre corren por el camino del PC. Solo Windows por ahora,
// que es la única plataforma que soporta esta app (ver foreground.js).
const POWER_COMMANDS = {
  sleep:     { win32: 'rundll32.exe powrprof.dll,SetSuspendState 0,1,0' },
  hibernate: { win32: 'shutdown /h' },
  restart:   { win32: 'shutdown /r /t 0' },
  shutdown:  { win32: 'shutdown /s /t 0' },
  lock:      { win32: 'rundll32.exe user32.dll,LockWorkStation' },
  logoff:    { win32: 'shutdown /l' },
};

function runPowerAction(mode) {
  const cmd = POWER_COMMANDS[mode]?.[process.platform];
  if (!cmd) {
    console.error(`Acción de energía "${mode}" no soportada en ${process.platform}`);
    return;
  }
  exec(cmd, (err) => {
    if (err) console.error(`Acción de energía "${mode}":`, err.message);
  });
}

// Traduce el mismo usage HID (Usage Page 0x07) que usa el editor de atajos
// (ver src/hid-keys.js, KEY_GROUPS) al nombre de tecla de nut.js. Cubre
// exactamente los códigos que ofrece el selector de teclas de la app.
const HID_TO_NUT_KEY = {
  // Letras 0x04-0x1D: A-Z
  ...Object.fromEntries(Array.from({ length: 26 }, (_, i) => [0x04 + i, String.fromCharCode(65 + i)])),
  // Números (fila superior) 0x1e-0x26: 1-9, 0x27: 0
  0x1e: 'Num1', 0x1f: 'Num2', 0x20: 'Num3', 0x21: 'Num4', 0x22: 'Num5',
  0x23: 'Num6', 0x24: 'Num7', 0x25: 'Num8', 0x26: 'Num9', 0x27: 'Num0',
  // F1-F12 (0x3a-0x45), F13-F24 (0x68-0x73)
  ...Object.fromEntries(Array.from({ length: 12 }, (_, i) => [0x3a + i, `F${i + 1}`])),
  ...Object.fromEntries(Array.from({ length: 12 }, (_, i) => [0x68 + i, `F${i + 13}`])),
  // Edición
  0x28: 'Return', 0x29: 'Escape', 0x2a: 'Backspace', 0x2b: 'Tab', 0x2c: 'Space',
  0x49: 'Insert', 0x4c: 'Delete', 0x4a: 'Home', 0x4d: 'End',
  0x4b: 'PageUp', 0x4e: 'PageDown', 0x46: 'Print',
  // Flechas
  0x50: 'Left', 0x4f: 'Right', 0x52: 'Up', 0x51: 'Down',
  // Símbolos
  0x2d: 'Minus', 0x2e: 'Equal', 0x2f: 'LeftBracket', 0x30: 'RightBracket',
  0x31: 'Backslash', 0x33: 'Semicolon', 0x34: 'Quote', 0x35: 'Grave',
  0x36: 'Comma', 0x37: 'Period', 0x38: 'Slash',
  // Teclado numérico
  0x54: 'Divide', 0x55: 'Multiply', 0x56: 'Subtract', 0x57: 'Add', 0x58: 'Enter',
  0x59: 'NumPad1', 0x5a: 'NumPad2', 0x5b: 'NumPad3', 0x5c: 'NumPad4', 0x5d: 'NumPad5',
  0x5e: 'NumPad6', 0x5f: 'NumPad7', 0x60: 'NumPad8', 0x61: 'NumPad9', 0x62: 'NumPad0',
};

// Bits de src/hid-keys.js MODIFIERS.
const HID_MODIFIER_TO_NUT_KEY = {
  0x01: 'LeftControl', 0x02: 'LeftShift', 0x04: 'LeftAlt', 0x08: 'LeftSuper',
  0x10: 'RightControl', 0x20: 'RightShift', 0x40: 'RightAlt', 0x80: 'RightSuper',
};

function modifierKeys(modifier, Key) {
  const keys = [];
  for (const [bit, name] of Object.entries(HID_MODIFIER_TO_NUT_KEY)) {
    if (modifier & Number(bit)) keys.push(Key[name]);
  }
  return keys;
}

// Hueco por defecto entre repeticiones cuando la acción no trae el suyo
// propio (secuencias guardadas antes de que el editor dejara elegirlo, ver
// DEFAULT_REPEAT_GAP_MS en profiles.js — debe coincidir con ese valor).
const DEFAULT_REPEAT_GAP_MS = 20;

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, Math.max(0, ms)));

// Hueco entre cada letra al escribir un texto letra a letra. Con 0 (lo que deja
// puesto loadNutJs para el resto de pasos) hay aplicaciones que se comen
// caracteres sueltos porque no llegan a procesar el evento anterior.
const TEXT_CHAR_DELAY_MS = 4;

// Escribir letra a letra tiene un techo que ningún ajuste arregla: cada carácter
// es un evento de teclado que el sistema tiene que entregar y la app destino
// procesar, así que un texto largo (una firma, un correo, un trozo de código) se
// ve teclear. A partir de este umbral se va por el portapapeles y un Ctrl+V, que
// mete el texto entero de golpe sin importar la longitud. Por debajo no
// compensa: pisar el portapapeles del usuario por cuatro letras sale más caro
// que escribirlas.
const TEXT_PASTE_MIN_CHARS = 5;

// Margen para que el portapapeles quede escrito antes de mandar el Ctrl+V, y
// para que la app destino lo lea antes de devolverle al usuario lo que tenía.
// Restaurarlo de inmediato hace que algunas apps peguen lo viejo: leen el
// portapapeles al procesar el atajo, no al recibirlo.
const CLIPBOARD_SETTLE_MS = 30;
const CLIPBOARD_RESTORE_MS = 120;

// Mete el texto de golpe por el portapapeles. Devuelve false si algo falla, para
// que el que llama pueda caer a escribirlo letra a letra.
//
// Se restaura el portapapeles anterior, pero solo su parte de texto: si lo que
// había copiado era una imagen o un archivo, se pierde. Y hay entornos donde
// Ctrl+V no pega (la consola clásica de Windows, algún juego, escritorios
// remotos); ahí el texto no aparece y no hay forma de detectarlo desde aquí.
async function pasteText(texto, keyboard, Key) {
  let previo = '';
  try {
    previo = clipboard.readText();
    clipboard.writeText(texto);
    await sleep(CLIPBOARD_SETTLE_MS);
    await keyboard.pressKey(Key.LeftControl, Key.V);
    await keyboard.releaseKey(Key.LeftControl, Key.V);
    await sleep(CLIPBOARD_RESTORE_MS);
    return true;
  } catch (err) {
    console.error('No se pudo pegar el texto por el portapapeles:', err.message);
    return false;
  } finally {
    if (previo) clipboard.writeText(previo);
  }
}

// Escribe un texto tal cual. No pasa por códigos de tecla HID (como sí hace
// "hotkey", ver más abajo), sino por el camino unicode del sistema: así sale
// igual con cualquier distribución de teclado -ñ, acentos, símbolos- sin tener
// que saber cuál tiene puesta el usuario. Eso mismo es lo que impide subirlo al
// teclado, que solo sabe mandar usages HID (ver DEVICE_STEP_TYPE en profiles.js).
//
// Un salto de línea o un tabulador no son caracteres que se puedan "escribir" ni
// pegar sueltos: se mandan como pulsación de Intro y de Tab, que es lo que el
// usuario espera (en un chat, Intro envía; pegar un "\n" no).
async function typeText(text, keyboard, Key) {
  if (!text) return;
  const previo = keyboard.config.autoDelayMs;
  keyboard.config.autoDelayMs = TEXT_CHAR_DELAY_MS;
  try {
    for (const trozo of String(text).split(/(\r\n|\n|\r|\t)/)) {
      if (!trozo) continue;
      if (trozo === '\t') {
        await keyboard.pressKey(Key.Tab);
        await keyboard.releaseKey(Key.Tab);
      } else if (trozo === '\n' || trozo === '\r' || trozo === '\r\n') {
        await keyboard.pressKey(Key.Return);
        await keyboard.releaseKey(Key.Return);
      } else if (trozo.length >= TEXT_PASTE_MIN_CHARS && await pasteText(trozo, keyboard, Key)) {
        continue;
      } else {
        await keyboard.type(trozo);
      }
    }
  } finally {
    keyboard.config.autoDelayMs = previo;
  }
}

async function runAction(action) {
  // Antes de cargar nut.js: un complemento no tiene por qué necesitar el ratón
  // ni el teclado (el de la lámpara solo usa la red), y sin esta salida
  // temprana un fallo al cargar el binario nativo (OneDrive con node_modules
  // como marcador de nube, por ejemplo) dejaría también sin funcionar lo que no
  // dependía de él.
  if (action.type === 'plugin') {
    await plugins.runStep(action);
    return;
  }

  const nut = loadNutJs();
  if (!nut) return;
  const { mouse, keyboard, screen, Point, Key, Button } = nut;

  if (action.type === 'center_mouse') {
    // Formato antiguo, de antes de poder capturar cualquier posición: se deja
    // por compatibilidad con secuencias ya guardadas.
    const [w, h] = await Promise.all([screen.width(), screen.height()]);
    await mouse.setPosition(new Point(w / 2, h / 2));
  } else if (action.type === 'mouse_position') {
    await mouse.setPosition(new Point(action.x, action.y));
  } else if (action.type === 'mouse_click') {
    const button = Button[CLICK_BUTTONS[action.button] || 'LEFT'];
    await repeatWithGap(action.count, action.gap, () => mouse.click(button));
  } else if (action.type === 'delay') {
    await sleep(action.ms || 0);
  } else if (action.type === 'key') {
    const key = mapCodeToNutKey(action.code, Key);
    if (key === null || key === undefined) return;
    await repeatWithGap(action.count, action.gap, async () => {
      await keyboard.pressKey(key);
      await keyboard.releaseKey(key);
    });
  } else if (action.type === 'text') {
    await repeatWithGap(action.count, action.gap, () => typeText(action.text, keyboard, Key));
  } else if (action.type === 'open_app') {
    // No hay opcode de firmware para esto (abrir algo requiere el SO): siempre
    // corre por el camino del PC, nunca se sube al teclado (ver DEVICE_STEP_TYPE
    // en profiles.js, que no lo incluye a propósito).
    if (!action.target) return;
    const err = await shell.openPath(action.target);
    if (err) console.error(`No se pudo abrir "${action.target}":`, err);
  } else if (action.type === 'system_power') {
    runPowerAction(action.mode);
  } else if (action.type === 'hotkey') {
    // Cualquier tecla (con o sin modificadores) del mismo selector que usa la
    // pestaña Atajo: { modifier, keycode } en usage HID.
    const mods = modifierKeys(action.modifier || 0, Key);
    const mainName = HID_TO_NUT_KEY[action.keycode];
    const main = mainName ? Key[mainName] : undefined;
    const keys = main !== undefined ? [...mods, main] : mods;
    if (!keys.length) return;
    await repeatWithGap(action.count, action.gap, async () => {
      await keyboard.pressKey(...keys);
      await keyboard.releaseKey(...keys);
    });
  }
}

// Repite un gesto (pulsación o clic) `count` veces, con un hueco de `gap` ms
// entre cada una: sin él, el SO puede llegar a fundir pulsaciones consecutivas
// en una sola (o no reconocer un doble clic como tal). Con `count` 1 o sin
// definir corre una sola vez, igual que antes de que existiera esta opción.
async function repeatWithGap(count, gap, run) {
  const times = Math.max(1, Math.round(count) || 1);
  const ms = Math.max(0, Math.round(gap ?? DEFAULT_REPEAT_GAP_MS));
  for (let i = 0; i < times; i++) {
    await run();
    if (i < times - 1) await sleep(ms);
  }
}

async function executeMacro(macroId) {
  const macro = (config.load().macros || []).find((m) => m.id === macroId);
  if (!macro) return;

  for (const action of macro.actions || []) {
    try {
      await runAction(action);
    } catch (err) {
      console.error(`Macro ${macroId}, acción "${action.type}":`, err.message);
    }
  }
}

// Posición del cursor en el mismo sistema de coordenadas que usa mouse.setPosition.
// Antes se leía con el módulo screen de Electron, pero en pantallas con escalado
// (125%/150% de Windows) ese sistema no coincide con el de nut.js y las posiciones
// capturadas quedaban con un desplazamiento al reproducirlas. Preguntándoselo a
// nut.js directamente, captura y reproducción siempre hablan el mismo idioma.
async function getMousePosition() {
  const nut = loadNutJs();
  if (!nut) return { x: 0, y: 0 };
  const p = await nut.mouse.getPosition();
  return { x: Math.round(p.x), y: Math.round(p.y) };
}

// Carga nut.js por adelantado al arrancar la app, en vez de esperar a la
// primera secuencia que caiga por el camino del PC (una macro no subida
// todavía al teclado, o con posición absoluta). Sin esto, esa primera
// ejecución paga sola el coste de cargar el binario nativo -normal que tarde
// notablemente, y peor aún si OneDrive tiene el node_modules como marcador de
// nube y tiene que bajarlo primero-, lo que se nota como un retraso enorme e
// inexplicable la primera vez.
function warmup() {
  loadNutJs();
}

module.exports = { executeMacro, getMousePosition, warmup };

File: KEYBOARD_ORBY_V3/RASPBERRY_PROJECT/ORBY_V4/OrbyGUI/src-tauri/src/macros.rs
// Las secuencias que ejecuta el PC.
//
// El teclado hace él solo lo que le cabe (retardos, atajos, clics), y manda `MACRO:<id>`
// solo cuando la secuencia tiene algún paso que no sabe hacer: una posición absoluta de
// ratón, un texto, abrir un programa, apagar el equipo o un complemento.
//
// La traducción de códigos de tecla vive en `orby_core::teclas`, que es donde se puede
// probar. Aquí solo está el trato con Windows.

use std::os::windows::process::CommandExt;
use std::sync::{Mutex, OnceLock};
use std::thread;
use std::time::Duration;

use orby_core::teclas::{code_a_vk, hid_a_vk, modificadores_a_vk, Tecla};
use serde_json::Value;
use tauri::AppHandle;
use tauri_plugin_clipboard_manager::ClipboardExt;
use windows::core::PCWSTR;
use windows::Win32::UI::Input::KeyboardAndMouse::{
    SendInput, INPUT, INPUT_0, INPUT_KEYBOARD, INPUT_MOUSE, KEYBDINPUT,
    KEYBD_EVENT_FLAGS, KEYEVENTF_EXTENDEDKEY, KEYEVENTF_KEYUP, KEYEVENTF_UNICODE, MOUSEINPUT,
    MOUSEEVENTF_LEFTDOWN, MOUSEEVENTF_LEFTUP, MOUSEEVENTF_MIDDLEDOWN, MOUSEEVENTF_MIDDLEUP,
    MOUSEEVENTF_RIGHTDOWN, MOUSEEVENTF_RIGHTUP, MOUSE_EVENT_FLAGS, VIRTUAL_KEY,
};
use windows::Win32::UI::Shell::ShellExecuteW;
use windows::Win32::UI::WindowsAndMessaging::{
    GetSystemMetrics, SetCursorPos, SM_CXSCREEN, SM_CYSCREEN, SW_SHOWNORMAL,
};

/// Hueco entre repeticiones cuando la acción no trae el suyo. Tiene que coincidir con el
/// `DEFAULT_REPEAT_GAP_MS` del editor (`src/views/profiles.js`).
const HUECO_POR_DEFECTO_MS: u64 = 20;

/// Con 0 hay aplicaciones que se comen letras sueltas porque no llegan a procesar el
/// evento anterior.
const RETARDO_ENTRE_LETRAS_MS: u64 = 4;

/// Por debajo de esto no compensa pisar el portapapeles del usuario. Por encima, escribir
/// letra a letra se ve teclear y no hay ajuste que lo arregle: cada carácter es un evento
/// que el sistema entrega y la app destino procesa.
const MINIMO_PARA_PEGAR: usize = 5;

/// Margen para que el portapapeles quede escrito antes del Ctrl+V, y para que la app
/// destino lo lea antes de devolverle al usuario lo que tenía. Restaurarlo de inmediato
/// hace que algunas apps peguen lo viejo: leen el portapapeles al procesar el atajo, no al
/// recibirlo.
const PORTAPAPELES_ASENTAR_MS: u64 = 30;
const PORTAPAPELES_RESTAURAR_MS: u64 = 120;

const VK_CONTROL_IZQ: u16 = 0xa2;
const VK_V: u16 = 0x56;
const VK_TAB: u16 = 0x09;
const VK_INTRO: u16 = 0x0d;

/// Dos secuencias pueden solaparse —el teclado no espera a que acabe una para mandar la
/// siguiente, y en Electron tampoco se esperaba—, pero el portapapeles es uno solo: sin
/// esto, la segunda lo pisa mientras la primera todavía no ha pegado y el usuario acaba
/// con el texto equivocado y el portapapeles de otro.
fn cerrojo_portapapeles() -> &'static Mutex<()> {
    static CERROJO: OnceLock<Mutex<()>> = OnceLock::new();
    CERROJO.get_or_init(|| Mutex::new(()))
}

fn dormir(ms: u64) {
    if ms > 0 {
        thread::sleep(Duration::from_millis(ms));
    }
}

// --- Envío de eventos a Windows ---

fn enviar(entradas: &[INPUT]) {
    unsafe { SendInput(entradas, std::mem::size_of::<INPUT>() as i32) };
}

fn evento_tecla(tecla: Tecla, soltar: bool) -> INPUT {
    let mut flags = KEYBD_EVENT_FLAGS(0);
    if tecla.extendida {
        flags |= KEYEVENTF_EXTENDEDKEY;
    }
    if soltar {
        flags |= KEYEVENTF_KEYUP;
    }

    INPUT {
        r#type: INPUT_KEYBOARD,
        Anonymous: INPUT_0 {
            ki: KEYBDINPUT {
                wVk: VIRTUAL_KEY(tecla.vk),
                wScan: 0,
                dwFlags: flags,
                time: 0,
                dwExtraInfo: 0,
            },
        },
    }
}

/// Un carácter suelto, por código Unicode en vez de por tecla virtual: así se escribe
/// igual sea cual sea la distribución del teclado del usuario, que es justo lo que
/// rompía escribir acentos con la vía de las teclas virtuales.
fn evento_unicode(unidad: u16, soltar: bool) -> INPUT {
    let mut flags = KEYEVENTF_UNICODE;
    if soltar {
        flags |= KEYEVENTF_KEYUP;
    }

    INPUT {
        r#type: INPUT_KEYBOARD,
        Anonymous: INPUT_0 {
            ki: KEYBDINPUT {
                wVk: VIRTUAL_KEY(0),
                wScan: unidad,
                dwFlags: flags,
                time: 0,
                dwExtraInfo: 0,
            },
        },
    }
}

fn evento_raton(flags: MOUSE_EVENT_FLAGS) -> INPUT {
    INPUT {
        r#type: INPUT_MOUSE,
        Anonymous: INPUT_0 {
            mi: MOUSEINPUT {
                dx: 0,
                dy: 0,
                mouseData: 0,
                dwFlags: flags,
                time: 0,
                dwExtraInfo: 0,
            },
        },
    }
}

fn pulsar(teclas: &[Tecla]) {
    let bajada: Vec<INPUT> = teclas.iter().map(|t| evento_tecla(*t, false)).collect();
    // Se sueltan en orden inverso: con Ctrl+Alt+B, soltar el Control antes que la B deja
    // a la aplicación destino viendo una B suelta.
    let subida: Vec<INPUT> = teclas.iter().rev().map(|t| evento_tecla(*t, true)).collect();
    enviar(&bajada);
    enviar(&subida);
}

fn escribir_caracter(c: char) {
    let mut buffer = [0u16; 2];
    for unidad in c.encode_utf16(&mut buffer) {
        enviar(&[evento_unicode(*unidad, false), evento_unicode(*unidad, true)]);
    }
}

// --- Pasos ---

fn repetir(accion: &Value, mut paso: impl FnMut()) {
    let veces = accion
        .get("count")
        .and_then(Value::as_f64)
        .map(|v| v.round() as i64)
        .filter(|v| *v >= 1)
        .unwrap_or(1);

    let hueco = accion
        .get("gap")
        .and_then(Value::as_f64)
        .map(|v| v.round().max(0.0) as u64)
        .unwrap_or(HUECO_POR_DEFECTO_MS);

    for i in 0..veces {
        paso();
        // Entre repeticiones, no después de la última: si no, cada secuencia arrastraría
        // un hueco de más al acabar.
        if i < veces - 1 {
            dormir(hueco);
        }
    }
}

/// Mete el texto de golpe por el portapapeles. `false` si algo falla, para que el que
/// llama caiga a escribirlo letra a letra.
///
/// Se restaura solo la parte de texto de lo que hubiera: si era una imagen o un archivo,
/// se pierde. Es lo mismo que hacía Electron.
fn pegar(app: &AppHandle, texto: &str) -> bool {
    let _guarda = cerrojo_portapapeles().lock().unwrap_or_else(|e| e.into_inner());

    let previo = app.clipboard().read_text().unwrap_or_default();

    if app.clipboard().write_text(texto.to_string()).is_err() {
        return false;
    }
    dormir(PORTAPAPELES_ASENTAR_MS);

    pulsar(&[
        Tecla { vk: VK_CONTROL_IZQ, extendida: false },
        Tecla { vk: VK_V, extendida: false },
    ]);

    dormir(PORTAPAPELES_RESTAURAR_MS);

    // La restauración va pase lo que pase por encima: dejar al usuario sin su portapapeles
    // es de los fallos que más molestan y que nadie asocia con el teclado.
    if !previo.is_empty() {
        let _ = app.clipboard().write_text(previo);
    }
    true
}

/// Un trozo de texto sin saltos: de golpe por el portapapeles si es largo, letra a letra
/// si es corto o si el pegado falla.
fn volcar(app: &AppHandle, trozo: &str) {
    if trozo.is_empty() {
        return;
    }
    if trozo.chars().count() >= MINIMO_PARA_PEGAR && pegar(app, trozo) {
        return;
    }
    for c in trozo.chars() {
        escribir_caracter(c);
        dormir(RETARDO_ENTRE_LETRAS_MS);
    }
}

fn escribir_texto(app: &AppHandle, texto: &str) {
    // Los saltos de línea y los tabuladores van como pulsaciones de verdad: pegados o
    // escritos como carácter, muchas aplicaciones no los interpretan.
    let mut trozo = String::new();
    let mut restantes = texto.chars().peekable();

    while let Some(c) = restantes.next() {
        match c {
            '\t' => {
                volcar(app, &trozo);
                trozo.clear();
                pulsar(&[Tecla { vk: VK_TAB, extendida: false }]);
            }
            '\n' | '\r' => {
                volcar(app, &trozo);
                trozo.clear();
                // «\r\n» es un salto, no dos.
                if c == '\r' && restantes.peek() == Some(&'\n') {
                    restantes.next();
                }
                pulsar(&[Tecla { vk: VK_INTRO, extendida: false }]);
            }
            _ => trozo.push(c),
        }
    }

    volcar(app, &trozo);
}

fn accion_energia(modo: &str) {
    // Sin opcode de firmware para esto: siempre corre por el camino del PC.
    let comando = match modo {
        "sleep" => "rundll32.exe powrprof.dll,SetSuspendState 0,1,0",
        "hibernate" => "shutdown /h",
        "restart" => "shutdown /r /t 0",
        "shutdown" => "shutdown /s /t 0",
        "lock" => "rundll32.exe user32.dll,LockWorkStation",
        "logoff" => "shutdown /l",
        _ => {
            eprintln!("[secuencias] acción de energía «{modo}» desconocida");
            return;
        }
    };

    // CREATE_NO_WINDOW: sin esto asoma una consola negra un instante cada vez que se
    // bloquea el equipo desde una tecla.
    const CREATE_NO_WINDOW: u32 = 0x0800_0000;
    if let Err(e) = std::process::Command::new("cmd")
        .args(["/C", comando])
        .creation_flags(CREATE_NO_WINDOW)
        .spawn()
    {
        eprintln!("[secuencias] acción de energía «{modo}»: {e}");
    }
}

fn abrir(destino: &str) {
    let ancho: Vec<u16> = destino.encode_utf16().chain(std::iter::once(0)).collect();
    let verbo: Vec<u16> = "open".encode_utf16().chain(std::iter::once(0)).collect();

    // Equivale al `shell.openPath` de Electron: abre con el programa asociado, valga lo
    // que valga el destino (un .exe, un documento, una carpeta o una URL).
    let r = unsafe {
        ShellExecuteW(
            None,
            PCWSTR(verbo.as_ptr()),
            PCWSTR(ancho.as_ptr()),
            None,
            None,
            SW_SHOWNORMAL,
        )
    };
    // Devuelve un pseudo-HINSTANCE: por debajo de 32 es un código de error.
    if r.0 as usize <= 32 {
        eprintln!("[secuencias] no se pudo abrir «{destino}»");
    }
}

fn ejecutar_accion(app: &AppHandle, accion: &Value) {
    let tipo = accion.get("type").and_then(Value::as_str).unwrap_or("");

    match tipo {
        "delay" => dormir(accion.get("ms").and_then(Value::as_u64).unwrap_or(0)),

        "mouse_position" => {
            let (Some(x), Some(y)) = (
                accion.get("x").and_then(Value::as_i64),
                accion.get("y").and_then(Value::as_i64),
            ) else {
                return;
            };
            // Coordenadas físicas del escritorio virtual, las mismas que devuelve
            // `mouse_get_position`. Ver el comentario de mouse.rs.
            let _ = unsafe { SetCursorPos(x as i32, y as i32) };
        }

        "center_mouse" => {
            // Formato antiguo: el centro de la pantalla principal.
            let ancho = unsafe { GetSystemMetrics(SM_CXSCREEN) };
            let alto = unsafe { GetSystemMetrics(SM_CYSCREEN) };
            let _ = unsafe { SetCursorPos(ancho / 2, alto / 2) };
        }

        "mouse_click" => {
            let (abajo, arriba) = match accion.get("button").and_then(Value::as_str) {
                Some("right") => (MOUSEEVENTF_RIGHTDOWN, MOUSEEVENTF_RIGHTUP),
                Some("middle") => (MOUSEEVENTF_MIDDLEDOWN, MOUSEEVENTF_MIDDLEUP),
                _ => (MOUSEEVENTF_LEFTDOWN, MOUSEEVENTF_LEFTUP),
            };
            repetir(accion, || {
                enviar(&[evento_raton(abajo), evento_raton(arriba)]);
            });
        }

        "key" => {
            let Some(tecla) = accion
                .get("code")
                .and_then(Value::as_str)
                .and_then(code_a_vk)
            else {
                return;
            };
            repetir(accion, || pulsar(&[tecla]));
        }

        "hotkey" => {
            let modificador = accion.get("modifier").and_then(Value::as_u64).unwrap_or(0) as u8;
            let mut teclas = modificadores_a_vk(modificador);
            if let Some(principal) = accion
                .get("keycode")
                .and_then(Value::as_u64)
                .and_then(|k| hid_a_vk(k as u8))
            {
                teclas.push(principal);
            }
            if teclas.is_empty() {
                return;
            }
            repetir(accion, || pulsar(&teclas));
        }

        "text" => {
            let Some(texto) = accion.get("text").and_then(Value::as_str) else { return };
            if texto.is_empty() {
                return;
            }
            repetir(accion, || escribir_texto(app, texto));
        }

        "open_app" => {
            if let Some(destino) = accion.get("target").and_then(Value::as_str) {
                if !destino.is_empty() {
                    abrir(destino);
                }
            }
        }

        "system_power" => {
            if let Some(modo) = accion.get("mode").and_then(Value::as_str) {
                accion_energia(modo);
            }
        }

        // Los complementos llegan en la Tarea 10. Hasta entonces se deja constancia en vez
        // de fallar en silencio, que es lo que haría creer que la tecla no está asignada.
        "plugin" => eprintln!("[secuencias] paso de complemento todavía sin portar"),

        // Un tipo desconocido no hace nada, igual que en Electron: una secuencia guardada
        // por una versión más nueva no puede tumbar a la vieja.
        otro => eprintln!("[secuencias] paso «{otro}» desconocido"),
    }
}

/// Mira si la línea es un disparo y, si lo es, ejecuta la secuencia. Se llama desde el
/// hilo del puerto, así que el trabajo se va a otro hilo: una secuencia con un `delay` de
/// diez segundos dejaría de leer el puerto y el vigilante daría la conexión por muerta.
pub fn quizas_disparar(app: &AppHandle, linea: &str) {
    let Some(id) = orby_core::protocolo::macro_trigger(linea) else { return };

    let app = app.clone();
    thread::spawn(move || ejecutar(&app, id));
}

fn ejecutar(app: &AppHandle, id: u32) {
    let config = crate::config::leer();
    let Some(macro_) = config
        .get("macros")
        .and_then(Value::as_array)
        .and_then(|lista| {
            lista
                .iter()
                .find(|m| m.get("id").and_then(Value::as_u64) == Some(id as u64))
        })
    else {
        eprintln!("[secuencias] disparo de la {id}, que no existe en la configuración");
        return;
    };

    // Las grabaciones las lleva recorder.rs (Tarea 7), no este ejecutor.
    match macro_.get("kind").and_then(Value::as_str) {
        Some("recording") | Some("recording-reset") => {
            eprintln!("[secuencias] la {id} es una grabación: todavía sin portar");
            return;
        }
        _ => {}
    }

    let Some(acciones) = macro_.get("actions").and_then(Value::as_array) else { return };

    // Un paso que falla no aborta la secuencia: si el sexto de ocho no se puede hacer, los
    // otros siete siguen valiendo.
    for accion in acciones {
        ejecutar_accion(app, accion);
    }
}
